src/main/scala/ir/Block.scala

```diff
@@ -0,0 +1,86 @@
+package ir
+
+import collection.mutable
+import scala.collection.immutable
+import intrusiveList.IntrusiveList
+import intrusiveList.IntrusiveListElement
+
+import scala.collection.mutable.ArrayBuffer
+
+class Block private 
+           (var label: String,
+             var address: Option[Int],
+             val statements: IntrusiveList[Statement],
+             // invariant: all Goto targets are disjoint 
+             private val _jumps: IntrusiveList[Jump],
+             //private val _calls: IntrusiveList[Call],
+             val incomingJumps: mutable.HashSet[Block],
+             val parent: Procedure
+           ) extends IntrusiveListElement {
+  def this(label: String, address: Option[Int], statements: IterableOnce[Statement], jumps: IterableOnce[Jump], parent: Procedure) = {
+    this(label, address, IntrusiveList.from(statements), IntrusiveList.from(jumps), mutable.HashSet.empty, parent)
+  }
+
+  def jumps: immutable.List[Jump] = _jumps to immutable.List
+  
+  def addJump(j: Jump): Unit = {
+    j.parent = this
+    _jumps.append(j)
+  }
+
+  def predecessors: immutable.Set[Block] = incomingJumps to immutable.Set
+ 
+  def removeJump(j: Jump) : Unit = {
+    assert(j.parent == this)
+    j.deParent()
+    _jumps.remove(j)
+  }
+
+  def replaceJump(j: Jump, newJ: Jump) : Unit = {
+    assert((j.parent == this))
+    if (j eq newJ) {
+     // GoTo/Call is responsible for maintaining the CFG if it is modified in-place
+      return
+    }
+    (j, newJ) match {
+       case (g: GoTo, f: GoTo) => {
+         removeJump(j)
+         addJump(newJ)
+       }
+       case (g: Call, f: Call) => {
+         removeJump(g)
+         addJump(f)
+       }
+       case (_, _) => throw Exception("Programmer error: can not replace jump with call or vice versa")
```

Comment:
Why is this not allowed? It is probably necessary for resolving jump tables, where an indirect call has to be resolved to gotos, not direct calls.